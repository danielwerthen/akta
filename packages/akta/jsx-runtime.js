export { Fragment, jsx, jsxs } from './dist';
