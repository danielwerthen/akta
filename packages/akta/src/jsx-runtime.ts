// Based on https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/314d0c3cf25e1762525a3790b4609eb19fddadf9/types/react/index.d.ts

import { AktaElement, AktaElementType } from './types';

export function jsx<
  P extends {},
  T extends string | ((props: P) => AktaElement<P, any>)
>(type: T, props: P, maybeKey?: string): AktaElement<P, T> {
  return {
    $$type: AktaElementType,
    type,
    props,
    key: maybeKey,
  };
}

export const jsxs = jsx;
