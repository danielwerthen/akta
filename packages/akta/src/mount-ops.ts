import {
  combineLatest,
  Observable,
  tap,
  isObservable,
  filter,
  switchMap,
  of,
  Subject,
} from 'rxjs';

type TempNode = null | string | TempNode[] | Observable<TempNode>;

export class Attacher {
  private live: boolean = false;
  private nodes: ChildNode[] = [];
  private sibling: () => ChildNode | null;
  private parent: () => ChildNode;
  private activator: Subject<void> | undefined;
  private activators: Subject<void>[] = [];
  constructor(
    sibling: () => ChildNode | null,
    parent: () => ChildNode,
    activator?: Subject<void>
  ) {
    this.sibling = sibling;
    this.parent = parent;
    this.activator = activator;
    activator?.subscribe(() => {
      this.activator = undefined;
    });
  }
  remove(idx: number) {
    if (this.live && this.nodes[idx]) {
      this.nodes[idx].remove();
    }
    delete this.nodes[idx];
  }
  attach(node: ChildNode, idx: number) {
    if (this.live && this.nodes[idx]) {
      this.nodes[idx].remove();
    }
    this.nodes[idx] = node;
    if (this.live) {
      const sib = this.nodes[idx - 1] ?? this.sibling();
      if (sib) {
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(node);
      } else {
        const p = this.parent();
        if (p.childNodes.length > 0) {
          p.firstChild?.before(node);
        } else {
          p.appendChild(node);
        }
      }
    }
  }
  private _activate() {
    this.live = true;
    for (let i = 0; i < this.nodes.length; i++) {
      const node = this.nodes[i];
      if (!node) {
        continue;
      }
      const sib = this.nodes[i - 1] ?? this.sibling();
      if (sib) {
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(node);
      } else {
        const p = this.parent();
        if (p.childNodes.length > 0) {
          p.firstChild?.before(node);
        } else {
          p.appendChild(node);
        }
      }
    }
    for (let i = 0; i < this.activators.length; i++) {
      this.activators[i].complete();
    }
  }
  activate() {
    if (this.activator) {
      this.activator.subscribe(() => this._activate());
    } else {
      this._activate();
    }
  }
  branch(idx: number): Attacher {
    const activator = new Subject<void>();
    this.activators.push(activator);
    return new Attacher(
      () => this.nodes[idx - 1],
      this.parent,
      this.live ? undefined : activator
    );
  }
  teardown() {
    let node: ChildNode | undefined;
    while ((node = this.nodes.pop())) {
      if (this.live) {
        node.remove();
      }
    }
  }
}

export function* attach(sibling: () => ChildNode) {
  let lineup: ChildNode[] = [];
  let active = false;
  let yieldData: null | (() => ChildNode) = null;
  while (true) {
    const action:
      | 'activate'
      | [ChildNode, number]
      | 'teardown'
      | number = yield yieldData;
    yieldData = null;
    if (action === 'activate') {
      if (active) {
        throw new Error('Activate can only be triggered once');
      }
      active = true;
      for (let i = 0; i < lineup.length; i++) {
        const sib = lineup[i - 1] ?? sibling();
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(lineup[i]);
      }
    } else if (Array.isArray(action)) {
      const [node, idx] = action;
      if (active && lineup[idx]) {
        lineup[idx].remove();
      }
      lineup[idx] = node;
      if (active) {
        const sib = lineup[idx - 1] ?? sibling();
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(node);
      }
    } else if (action === 'teardown') {
      for (let i = 0; i < lineup.length; i++) {
        lineup[i].remove();
      }
      break;
    } else if (typeof action === 'number') {
      const idx: number = action;
      yieldData = () => lineup[idx - 1] ?? sibling();
    }
  }
}
function onlyFirst(_value: unknown, idx: number) {
  return idx === 0;
}

export function attachChildren(
  attacher: Attacher,
  children: TempNode,
  index?: number
): Observable<unknown> | void {
  if (!children) {
    const node = document.createTextNode('');
    attacher.attach(node, index ?? 0);
  } else if (typeof children === 'string') {
    const node = document.createTextNode(children);
    attacher.attach(node, index ?? 0);
  } else if (Array.isArray(children)) {
    const branch =
      typeof index === 'number' ? attacher.branch(index) : attacher;
    const observables = children
      .map((child, idx) => {
        return attachChildren(branch, child, idx);
      })
      .filter<Observable<unknown>>(
        (item: unknown): item is Observable<unknown> => !!item
      );
    return new Observable(sub => {
      sub.add(() => branch.teardown());

      if (observables.length < 1) {
        if (typeof index === 'number') {
          attacher.remove(index);
        }
        branch.activate();
        sub.next();
        return;
      }
      return combineLatest(observables)
        .pipe(
          tap(() => {
            if (typeof index === 'number') {
              attacher.remove(index);
            }
            branch.activate();
            sub.next();
          })
        )
        .subscribe();
    });
  } else if (isObservable(children)) {
    return children.pipe(
      switchMap(child => {
        const obs = attachChildren(attacher, child, index);
        if (!obs) {
          return of(void 0);
        }
        return obs;
      }),
      filter(onlyFirst)
    );
  }
}
