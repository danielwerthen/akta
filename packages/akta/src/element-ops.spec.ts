import { Subject, firstValueFrom } from 'rxjs';
import { take, toArray } from 'rxjs/operators';
import { jsx } from './jsx-runtime';
import { mount } from './mount-ops';

describe('Element ops', () => {
  it('should handle observable classNames', async () => {
    const root = document.createElement('div');
    const subj = new Subject();
    const unsub = mount(jsx('div', { className: subj }), root);
    subj.next('cl454');
    expect(root).toMatchSnapshot();
    subj.next(3);
    expect(root).toMatchSnapshot();
    subj.next(null);
    expect(root).toMatchSnapshot();
    subj.next(undefined);
    expect(root).toMatchSnapshot();
    subj.next('final');
    expect(root).toMatchSnapshot();
    unsub();
  });
  it('should handle onclick events', async () => {
    const root = document.createElement('div');
    document.body.append(root);
    const subj = new Subject();
    const unsub = mount(jsx('button', { onClick: subj }), root);
    const res = firstValueFrom(subj.pipe(take(4), toArray()));
    const btn = document.getElementsByTagName('button')[0];
    btn.click();
    btn.click();
    btn.click();
    btn.click();
    const final = await res;
    expect(final).toMatchSnapshot();
    unsub();
    document.body.removeChild(root);
  });
  it('should handle onclick functions', async () => {
    const root = document.createElement('div');
    document.body.append(root);
    const fun = jest.fn();
    const unsub = mount(jsx('button', { onClick: fun }), root);
    const btn = document.getElementsByTagName('button')[0];
    btn.click();
    btn.click();
    btn.click();
    btn.click();
    expect(fun.mock.calls).toMatchSnapshot();
    unsub();
    document.body.removeChild(root);
  });
  it('should handle aria attributes', async () => {
    const root = document.createElement('div');
    const unsub = mount(jsx('button', { ['aria-hidden']: true }), root);
    expect(root).toMatchSnapshot();
    unsub();
  });
});
