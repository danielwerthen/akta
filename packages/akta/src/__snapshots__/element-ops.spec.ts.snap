// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element ops should handle aria attributes 1`] = `
<div>
  <button
    aria-hidden="true"
  />
</div>
`;

exports[`Element ops should handle observable classNames 1`] = `
<div>
  <div
    class="cl454"
  />
</div>
`;

exports[`Element ops should handle observable classNames 2`] = `
<div>
  <div
    class="3"
  />
</div>
`;

exports[`Element ops should handle observable classNames 3`] = `
<div>
  <div
    class=""
  />
</div>
`;

exports[`Element ops should handle observable classNames 4`] = `
<div>
  <div
    class=""
  />
</div>
`;

exports[`Element ops should handle observable classNames 5`] = `
<div>
  <div
    class="final"
  />
</div>
`;

exports[`Element ops should handle onclick events 1`] = `
Array [
  MouseEvent {
    "isTrusted": false,
  },
  MouseEvent {
    "isTrusted": false,
  },
  MouseEvent {
    "isTrusted": false,
  },
  MouseEvent {
    "isTrusted": false,
  },
]
`;

exports[`Element ops should handle onclick functions 1`] = `
Array [
  Array [
    MouseEvent {
      "isTrusted": false,
    },
  ],
  Array [
    MouseEvent {
      "isTrusted": false,
    },
  ],
  Array [
    MouseEvent {
      "isTrusted": false,
    },
  ],
  Array [
    MouseEvent {
      "isTrusted": false,
    },
  ],
]
`;
